[FLINK-24138] Architectural tests

Airblader opened the pull request:
## What is the purpose of the change

This pull request adds a new `flink-architecture-tests` module for architectural tests using [ArchUnit](https://www.archunit.org/). 

~It currently also contains two commits which address some of the violations as a demonstration, but we can separate those out if needed.~ Edit: These commits have been removed for now.

A big thanks goes out to @codecholeric who reviewed this and provided me with valuable feedback as well as his time for some questions.

## Verifying this change

This pull request adds all the tests. :-)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes (test-only)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

Airblader commented on flink-table/flink-table-common/src/test/java/org/apache/flink/table/architecture/ArchitectureTests.java:
```diff
@@ -0,0 +1,100 @@
+
+    /** Ensure that all classes in the API package have a proper annotation. */
+    @ArchTest
+    public static final ArchRule ANNOTATED_APIS =
```
In flink-table-common this test currently identifies a couple of classes (all exceptions) which are not annotated.

Airblader commented on flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/architecture/ApiRules.java:
```diff
@@ -0,0 +1,93 @@
+     * and argument types which are also marked {@link Public} or {@link PublicEvolving}.
+     */
+    @ArchTest
+    public static final ArchRule PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES =
```
In flink-table-common this test currently identifies quite a few problems, e.g. a lot of methods in `Catalog` referencing non-annotated classes, or `SinkProvider` being PublicEvolving, but `Sink` being Experimental.

Airblader commented on flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/architecture/Predicates.java:
```diff
@@ -0,0 +1,158 @@
+                                .collect(Collectors.joining(", "))) {
+            @Override
+            public boolean apply(JavaClass clazz) {
+                final JavaClass enclosingClazz = getEnclosingClass(clazz).getBaseComponentType();
```
This is a bit of a Flink specialty, I guess, since we decorate only the top-level classes, not nested classes directly.

Airblader commented on flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/architecture/ApiRules.java:
```diff
@@ -0,0 +1,93 @@
+     * and argument types which are also marked {@link Public} or {@link PublicEvolving}.
+     */
+    @ArchTest
+    public static final ArchRule PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES =
```
We could also think about making these stricter: instead of only checking argument/return types which reside in o.a.f, check any classes which do not reside in java.*. This might require some additional allow-list. It's a question of how much we want to allow potentially relying on classes which are neither from the SDK nor Flink.

sjwiesman replied:
Not a review, I was just peaking around at how this works. 

In general the only allow list is:
- org.apache.flink.* but _not_ org.apache.flink.shaded (!)
- java.*
- scala.* (only in the scala api's)
- Kryo https://github.com/apache/flink/blob/99c2a415e9eeefafacf70762b6f54070f7911ceb/flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java#L720-L812 

And then connectors will often expose some number of classes from their underlying client. For example, the Kafka serde classes expose ConsumerRecord and ProducerRecord.

Airblader replied:
Thanks, Seth, that sounds like a good list. I think shaded and Scala we would then consider directly in the rule; for Kryo and connector-specific things I would probably allow the violation and freeze them. Being aware and made aware on changes that further expose third-party classes would be valuable, and a connector-specific allow-list would be needed either way anyway.

zentol commented on flink-architecture-tests/src/test/java/org/apache/flink/architecture/common/Conditions.java:
```diff
@@ -0,0 +1,143 @@
+     *   <li>Otherwise, check just the type itself.
+     * </ul>
+     */
+    public static ArchCondition<JavaMethod> haveLeafReturnTypes(
```
is there a simple way to test these conditions/predicates/rules etc.?

Airblader replied:
I like to think that they are tested implicitly by the violations they produce (i.e. any change to the rule will inevitably be checked against the entire codebase and all frozen violations).

Testing should be fairly easy, though, because you can create any sample code you want in a package and run the rule only on that package and assert the violations. I'll have to look if they could also be unit-tested, which only relies on the ability to instantiate `JavaClass` etc. as mocks.

codecholeric replied:
Originally within ArchUnit I had also attempted to unit test these things with sort of stub imported objects, but eventually I switched over to simply creating test examples and using the real `ClassFileImporter` for imports. It's just more blackbox and the tests are still super fast. If you really want to mock-unit-test something like this without creating an example class you could ponder about making the signature

```
public static <T extends HasName.AndFullName & HasReturnType> ArchCondition<T> haveLeafReturnTypes(...)
```

Then you could simply mock the interface. But typically within ArchUnit we use tests with local classes where possible, like

```
@Test
void example() {
  class SomeExample {...}

  new ClassFileImporter().importClass(SomeExample.class);
}
```

codecholeric replied:
Or you create a little test samples project that contains all your typical violations that you would wanna find and make sure your rules catch those.

zentol reviewed (commented):
The current rule set is rather prohibitive when it comes to `@Experimental` APIs.

For example, this would be forbidden:

```
@Public
class X {

   @Experimental
   XExperiment doSomethingExperimental(XEperiment x){...}

   @Experimental // this could also be a top-level classes
   static class XExperiment {...}
}
```

This looks rather reasonable to me. As is the rules would reinforce the current pattern of marking everything as `@PublicEvolving`.

Airblader commented on flink-architecture-tests/src/test/resources/archunit.properties:
```diff
@@ -0,0 +1,24 @@
+# Path under which known violations will be stored for frozen rules
+freeze.store.default.path=violations
+
+# Enable this to update the violations files
```
I still want to add more documentation here, but I'm still playing around with the best setup. On the CI, changes to violations should always fail. For local execution we can decide between failing and automatically updating the violation store. The latter is more convenient, but has a slight risk that people just "accept" those changes. I'm not sure what the better approach is (yet).

Airblader replied:
Since there are separate settings for creating and updating violation stores, my preferred approach at the moment would be to automatically update stores locally, but not create them. On CI both are disabled. I think this reduces developer frustration a bit since resolving a violation would automatically be recorded and not require changing some config and running the test again.

zentol replied:
Is there no way to configure it to work as remove-only? If a violation is fixed, remove entry. If a new one is added, fail.

That would be the only "sane" approach in my opinion. If by default we just add a violation then no one will fix this stuff.

Airblader replied:
We could implement our own violation store, but in the default one that doesn't seem to be possible. I've raised https://github.com/TNG/ArchUnit/issues/670 for this for now. I don't think I'd start with a custom violation store for the time being and instead have a separate target which can be run to update violations, but disallow updates by default (local & CI).

Airblader replied:
Based on someone's feedback I actually verified that this already behaves the way we want to, i.e. when run locally new violations will lead to a failure, but removed violations will be updated. On the CI we fail for any change to the violation store, even removal, since that removal needs to be committed.

Airblader commented:
@zentol To reply to a bunch of your points at once: I think we should hit a "sweet spot" with the rules not being too complex. For example, if we allow Experimental methods in Public APIs(*), we could just accept that those will be marked as violations as over time the expectation has to be that it does change (either it becomes stable or gets removed again). WDYT?

(*) In fact I'm not sure we have a project-wide consensus on how these annotations are to be used. flink-core handles things very differently from flink-table, for example.

Airblader commented on flink-architecture-tests/src/test/java/org/apache/flink/architecture/rules/ApiAnnotationRules.java:
```diff
@@ -0,0 +1,168 @@
+                            .and()
+                            .arePublic()
+                            .and()
+                            .areNotAnnotatedWith(Internal.class)
```
Yes, that is a fair point. I haven't dug into all violations in detail yet, but I can check whether this changes the identified violations. If not, then there wouldn't be a reason to add this, since this scenario wouldn't be able to arise anymore.

zentol commented:
> we could just accept that those will be marked as violations as over time the expectation has to be that it does change (either it becomes stable or gets removed again). WDYT?

IMO expected/allowed changes should never result in a violation. If we allow experimental methods in public interface, then it should not be a violation. Otherwise the semantics of a violation becomes ambiguous.

Airblader commented:
@zentol In that case we probably need to more clearly define how annotations should be used in the project first as we'd very strictly want to forbid any new violations rather than use it as a guide and way to make possible unintentional violations visible. 

Currently, modules like flink-core and flink-table seem to have a very different understanding of how annotations should be used. It heavily annotates all nested classes, not just top-level ones. Also, would that mean that methods are allowed to be annotated with any less-stable annotation than their top-level class in general (including `@Internal`)?

zentol commented:
> [...] methods are allowed to be annotated with any less-stable annotation than their top-level class in general (including @Internal)?

Yes. There are no rules against that, and IIRC it is done extensively in the Dataset/Datastream APIs. If the Table API got by without it then that is great, but I wouldn't conclude that this implies some misunderstanding as to how the annotations may be used. I would rather attribute that to the fact that the Table API doesn't have any `@Public` APIs at this point, so they can change things much more freely anyway.

Airblader commented:
That makes a lot of sense. I can adjust the rules to reflect that. The question which remains, however, is whether annotating all (nested) classes or only top-level classes is the "correct" approach. Is there an answer to this or is this something we should clarify in the community?

zentol commented:
Nested classes are a special case and also need to be annotated separately; this is restriction imposed by japicmp.

Airblader commented:
Thanks! That's good to know, and clearly not how flink-table currently seems to do it. Would it make sense to enforce that _all_ (public?) classes (maybe except for a module-deny-list?) are required to have a visibility annotation?

zentol commented:
>  Would it make sense to enforce that all **public** classes are required to have a visibility annotation

I'm not sure. On the one hand it forces the dev to make a decision w.r.t. visibility, which is good, but at the same time I could imagine this being very annoying overhead. Because of that I'm worried that we'd just end up slapping `@PublicEvolving` on everything and calling it a day (which de-facto already happens in the APIs). And then I'm wondering what the real benefit actually is.

Actually, I would not enforce that for now, because we should rethink `@Experimental/@PublicEvolving` anyway. As is we enforce _nothing_, so really, what's the point.

Airblader commented:
> Because of that I'm worried that we'd just end up slapping @PublicEvolving on everything

While I understand what you're getting at, this is a human problem that tooling cannot solve. We have code reviews and merging is limited to committers – we need to be able to trust that such "abuse" or carelessness doesn't happen, or address it if it does. Ultimately, we need to be able to change APIs, meaning there will always be a way to make the CI happy, and that also means there's always a way for people to "just do" something to bypass these tools.

Tooling like this prevents changes to the API when they would happen accidentally / unknowingly. For a human it's essentially impossible to notice that a public API is returning an internal type somewhere down the line etc., and that's what the tooling can prevent.

> but at the same time I could imagine this being very annoying overhead

This is a bit besides the point, but I'd say that having to add JavaDocs to every public class is significantly more annoying since it requires actually writing custom text, even on test classes.

---

A whole different approach that I know from other projects (not from the Java ecosystem, though) is generating a documentation of all public APIs in the CI and also versioning that separate from the code; if someone changes an API (be it accidentally or intentionally), the CI will fail. Having to update this file in the PR makes it impossible to not be aware of the change to the API you're doing. But even then it's the committer's job to review this change and ensure it is OK.

Airblader commented:
FWIW, ensuring that all classes have such an annotation is something that, if at all, I'd do in Checkstyle because you really want that to be immediate feedback to reduce frustration.

Airblader commented:
I've rebased the PR and addressed most of the feedback; in particular, I've changed it to not check API annotations on the enclosing class anymore but on the classes directly. This of course produces a lot of violations, as do some other rules; many of them are immediate fixes, and I plan on doing a bunch of those directly in this PR.

There are also a lot of violations that need to be addressed properly. Once the list of violations is cut down a bit I might investigate and possibly open JIRA issues for them.

Once the CI passes I'll also intentionally push a change to verify that removing or adding a violation correctly fails in the CI.

Airblader commented:
OK, I hopefully have resolved the issues with Maven now, let's see what the CI says. Essentially when run in Maven, the dependencies come in as JARs (d'oh). But previously, I had the import option to exclude JARs enabled. Disabling that option in turn blew up memory usage and the whole thing OOM'ed. 
I now added an import option to exclude shaded stuff, and that seems to work fine even with default memory settings. Also, the current execution time for the tests is more on the order of a few seconds, not three minutes, so :tada: to that.

The PR now also contains two commits which resolve some of the violations that seemed trivial to resolve. There's plenty of non-trivial ones, and once (and if) we have this merged I'd like to go through more of them to either resolve them or create dedicated issues for some of the more tricky ones.

Airblader commented on flink-architecture-tests/violations/d208df3e-0594-4d6c-82bb-94b12f68ec1f:
```diff
@@ -0,0 +1,66 @@
+Constructor <org.apache.flink.runtime.scheduler.exceptionhistory.ExceptionHistoryEntry.<init>(java.lang.Throwable, long, java.lang.String, org.apache.flink.runtime.taskmanager.TaskManagerLocation)> calls method <org.apache.flink.runtime.scheduler.exceptionhistory.ExceptionHistoryEntry$ArchivedTaskManagerLocation.fromTaskManagerLocation(org.apache.flink.runtime.taskmanager.TaskManagerLocation)> in ...
```
Just as an FYI, the line numbers contained at the end here are ignored during violation matching, so we don't need to worry that this will constantly break because code moves around (within the same file).

Airblader commented on flink-architecture-tests/src/test/java/org/apache/flink/architecture/rules/ApiAnnotationRules.java:
```diff
@@ -0,0 +1,160 @@
+                            .and()
+                            .areNotAnnotatedWith(Internal.class)
+                            .and()
+                            .areNotAnnotatedWith(Deprecated.class)
```
I guess we should also exclude PublicEvolving and Experimental here.

Airblader commented on flink-architecture-tests/src/test/java/org/apache/flink/architecture/rules/ApiAnnotationRules.java:
```diff
@@ -0,0 +1,160 @@
+                            .and()
+                            .areNotAnnotatedWith(Internal.class)
+                            .and()
+                            .areNotAnnotatedWith(Deprecated.class)
```
Need to exclude Experimental here, too

Airblader commented on flink-architecture-tests/src/test/java/org/apache/flink/architecture/rules/ApiAnnotationRules.java:
```diff
@@ -0,0 +1,160 @@
+            freeze(
+                    methods()
+                            .that()
+                            .areDeclaredInClassesThat(
```
Maybe we can expand this to "or the method is annotated PublicEvolving"

zentol commented on flink-architecture-tests/src/test/java/org/apache/flink/architecture/rules/ApiAnnotationRules.java:
```diff
@@ -0,0 +1,168 @@
+            freeze(
+                    methods()
+                            .that()
+                            .areNotAnnotatedWith(PublicEvolving.class)
+                            .and()
+                            .areDeclaredInClassesThat(
+                                    areJavaClasses()
+                                            .and(
+                                                    areDirectlyAnnotatedWithAtLeastOneOf(
+                                                            Public.class, PublicEvolving.class)))
```
I don't quite recall our previous discussions, but this looks a bit odd. Why are we excluding methods that are annotated as PublicEvolving? Why do we also check classes that are Public (combined these should only cover Public methods)?

If I have a PublicEvolving class, and an explicit PublicEvolving method, then will it not be ignored? Similar case for a Public class and a PublicEvolving method.

I expected something like
`
if annotatedWith(PublicEvolving) || residesInClassWithAnnotation(PublicEvolving.class)
`

Airblader replied:
D'oh, you're absolutely right. Fixed.
